1.20.1/2.3 create6

chj990707 opened the pull request:
Rendering of create GhostBlockRenderer on vs ship is implemented.

- GhostBlockRenderer checks for BE position if it is on a ship and transforms itself.
- A bit rubber-bands when the ship is moving. Suspected if I called (tick) Transformation instead of RenderTransformation but I didn't. Might need polishing.

Rendering of flywheel BlockEntityVisuals on vs ship is implemented.

- Originally BlockEntityStorage class of flywheel holds BlockEntityVisuals for rendering, and they apply correct transformation to their Instances.
- To implement rendering on ship, BlockEntityStorage is injected to check if the newly added BE is on a ship and register the relation.
- BlockEntityStorage also creates & updates VisualEmbedding, subclass of VisualizationContext which is in the original flywheel API to support arbitrary transformation of multiple visuals, thereby creating a 'meta-visual'.
- BlockEntityStorage.vs$updateAllShips() should be called to update VisualEmbedding transformations, but Storage classes themselves don't have tick() or render() methods.
- Instead, it is called by VisualizationManagerImpl on every render() calls, and updates the transformation of each BlockEntityVisuals on ships.
- Changes made in f2b39f12 was reverted to remove the duplicate implementation, and confirmed the functionality.

NOTE: This PR intentionally excluded changes made in other files, which were just package renamings, disabling injections missing targets, and dependency version change. It won't ensure the build success.

chj990707 committed:
Updated annotation processor for mixin extras from 0.2.0 to 0.4.1
Fixed MixinGhostBlockRenderer. Instead of redirecting every call to translate, it will modify the PoseStack by Wrapping the method.

chj990707 committed:
Missed change from commit 5aa71ee13c56cf00dad823beb9cce2edb06f2052

chj990707 committed:
Flywheel BlockEntityVisuals compatibility implemented.
- BlockEntities registered to the flywheel BlockEntityStorage will be automatically checked if it was placed on a ship.
- If it is on a ship, then a VisualEmbedding attached to the ClientShip will be created and registered for reuse.
- VisualEmbedding transformations that are attached to ships are updated by VisualizationManager.render()
- Found out missed mixins.json line that was forgotten at previous commit(again) and added it.
- Removed classes that refers to previous structure of flywheel.

chj990707 committed:
Merge remote-tracking branch 'vs_team/1.20.1/2.3-create6' into 1.20.1/2.3-create6

# Conflicts:
#	common/src/main/java/org/valkyrienskies/mod/compat/FlywheelEvents.java
#	common/src/main/java/org/valkyrienskies/mod/mixin/mod_compat/create/client/MixinGhostBlockRenderer.java
#	common/src/main/resources/valkyrienskies-common.mixins.json

chj990707 committed:
Changes made in f2b39f12513219f5e9189fa1b2e84debf0758b09 reverted, and confirmed the visuals still rendering by changes made in 20c72e9b591587988868c3e045db4e0e9a14ce31

blockninja124 commented:
PRs changing the dependency number, fixing packages, and other changes are totally fine btw. The only thing I want to avoid (although it might still happen, from myself or PRs) is disabling stuff that isn't fixed purely to get a compilation.

But with how it is right now, even those changes might be better than nothing 🤔

chj990707 committed:
- MixinBlockEntityStorage has been refactored into MixinStorage, allowing generic over blockentity and entity.
- Contraptions will now render normally on ship, minus relighting problem.
- Fixed the bug of not being able to render visuals on ship properly when it was loaded from getting closer. MixinStorage will now track the renderOrigin of original VisualizationContext.
- MixinTileEntityRenderHelper has been removed from common and renamed to MixinBlockEntityRenderHelper in fabric/forge. Each has different method parameters.
- VisualizationManagerImpl dependency on updating MixinStorage was replaced by direct refer to MixinStorage, removing MixinVisualizationManagerImpl.
- Removed MixinOrientedContraptionEntity. Why is this class empty?

chj990707 committed:
ContraptionRenderInfo is removed from create fabric 6.0.7 snapshot.
MixinFlwContraption, which is targetting ContraptionMatrices, has been moved to common.

chj990707 committed:
Dependency range for forge has been changed.

chj990707 commented:
- Relying on VisualizationManagerImpl.render to update the ship transform to the embedding is moved to FlwEvents, removing potential danger of mixin to impl class.

chj990707 committed:
LightBox AABB checking was moved to MixinContraptionVisual.

chj990707 committed:
Removed unused packages

chj990707 committed:
Vanillin is added to gradle.
Visibility test for AbstractEntityVisual will always return true if the entity is on a ship.
MixinMinecartInstance is removed; Generic visual modification at MixinStorage will handle the relevant operations.
There is a bug that will load the entities on a ship before the ship.

chj990707 committed:
Renamed local variables in MixinStorage and cleaned up reused code for embedding transformation.

chj990707 committed:
Creating and managing visual embeddings for the ships is now moved from MixinStorage to MixinVisualizerRegistry and decorators of Visualizer classes.
Hopefully this will be more stable since it only relies on flywheel api classes.

chj990707 committed:
Entity Culling check has been moved from MixinAbstractEntityVisual to MixinEntityVisibilityTester, and actually checks if the entity is in frustum.
Updated TODO list.

chj990707 committed:
Removed FlywheelEvents class and moved its function to ShipEmbeddingManager.
Fixed event handlers not registering due to lazy intialization.
Added synchronized keyword to ShipEmbeddingManager methods.

chj990707 commented:
- Mixins to flywheel backend classes are removed. It now has 2 Mixins to API classes, each for decorating visualizer and Entity visibility check.
- Works with vanillin rendering.
- Visual Embedding operation is synchronized.

chj990707 committed:
Fixed visuals being rubber-banded when the ship is moving.
- update of visual embedding transform is moved to rendership event.
